src/core/annotation.js

```diff
@@ -282,6 +295,7 @@ class Annotation {
       modificationDate: this.modificationDate,
       rect: this.rectangle,
       subtype: params.subtype,
+      documentHasJSActions: params.hasJSActions,
```

Comment:
> I don't really see why we need to and would want to store a document-wide property on every individual annotation, especially since this is then also an API change. 

Sorry, but how is adding a *new* boolean to the data returned by `getAnnotations` a problem? :-)
The data in question is also cached, on the worker-side, so that shouldn't be much of an issue.

The reason why I suggested the previous approach, is that doing it the way you suggest (and what the patch originally did), has *really* undesirable consequences for the viewer since that first of all gives a number of the viewer-components (e.g. `PDFPageView`) *very* inconsistent API-surfaces!
Note also the point below, about unnecessary things API-calls running too early.

> It would be much better to use the same logic as enableScripting, which is also a document-wide property, where we calculate the property once and pass it into the AnnotationLayer class.

Sorry, but I don't agree with this at all, since it ends up forcing an unnecessary API-call early on, *before* rendering has started. That's why I suggested the previous approach, since it delays these things until the Annotations are actually fetched.
You absolutely want to avoid *all* strictly unnecessary API-calls *before* rendering has started!

---

Given that we apparently have very different positions/perspectives here, please give me some time to suggest a compromise here!

Reply:
I'd suggest the following approach, to ensure that `hasJSActions` isn't called *before* rendering has started and avoid `BaseViewer`/`PDFPageView` diverging unnecessarily:
```diff
diff --git a/src/display/annotation_layer.js b/src/display/annotation_layer.js
index 705568575..b8e1c90fe 100644
--- a/src/display/annotation_layer.js
+++ b/src/display/annotation_layer.js
@@ -43,6 +43,8 @@ import { AnnotationStorage } from "./annotation_storage.js";
  *   for annotation icons. Include trailing slash.
  * @property {boolean} renderInteractiveForms
  * @property {Object} svgFactory
+ * @property {boolean} [enableScripting]
+ * @property {boolean} [hasJSActions]
  */
 
 class AnnotationElementFactory {
diff --git a/src/display/api.js b/src/display/api.js
index 8ffcc1495..b58830463 100644
--- a/src/display/api.js
+++ b/src/display/api.js
@@ -904,8 +904,8 @@ class PDFDocumentProxy {
   }
 
   /**
-   * @returns {Promise<boolean>} A promise that is resolved with true
-   * if some fields have some JS actions.
+   * @returns {Promise<boolean>} A promise that is resolved with `true` if some
+   *   /AcroForm fields has JavaScript actions.
    */
   hasJSActions() {
     return this._transport.hasJSActions();
diff --git a/web/annotation_layer_builder.js b/web/annotation_layer_builder.js
index e37ee831f..c4d47112a 100644
--- a/web/annotation_layer_builder.js
+++ b/web/annotation_layer_builder.js
@@ -28,7 +28,8 @@ import { SimpleLinkService } from "./pdf_link_service.js";
  * @property {IPDFLinkService} linkService
  * @property {DownloadManager} downloadManager
  * @property {IL10n} l10n - Localization service.
- * @property {boolean} enableScripting
+ * @property {boolean} [enableScripting]
+ * @property {boolean} [hasJSActionsPromise]
  */
 
 class AnnotationLayerBuilder {
@@ -45,7 +46,7 @@ class AnnotationLayerBuilder {
     renderInteractiveForms = true,
     l10n = NullL10n,
     enableScripting = false,
-    hasJSActions = false,
+    hasJSActionsPromise = null,
   }) {
     this.pageDiv = pageDiv;
     this.pdfPage = pdfPage;
@@ -56,7 +57,7 @@ class AnnotationLayerBuilder {
     this.l10n = l10n;
     this.annotationStorage = annotationStorage;
     this.enableScripting = enableScripting;
-    this.hasJSActions = hasJSActions;
+    this._hasJSActionsPromise = hasJSActionsPromise;
 
     this.div = null;
     this._cancelled = false;
@@ -69,7 +70,10 @@ class AnnotationLayerBuilder {
    *   annotations is complete.
    */
   render(viewport, intent = "display") {
-    return this.pdfPage.getAnnotations({ intent }).then(annotations => {
+    return Promise.all([
+      this.pdfPage.getAnnotations({ intent }),
+      this._hasJSActionsPromise,
+    ]).then(([annotations, hasJSActions = false]) => {
       if (this._cancelled) {
         return;
       }
@@ -88,7 +92,7 @@ class AnnotationLayerBuilder {
         downloadManager: this.downloadManager,
         annotationStorage: this.annotationStorage,
         enableScripting: this.enableScripting,
-        hasJSActions: this.hasJSActions,
+        hasJSActions,
       };
 
       if (this.div) {
@@ -133,8 +137,8 @@ class DefaultAnnotationLayerFactory {
    *   for annotation icons. Include trailing slash.
    * @param {boolean} renderInteractiveForms
    * @param {IL10n} l10n
-   * @param {boolean} enableScripting
-   * @param {boolean} hasJSActions
+   * @param {boolean} [enableScripting]
+   * @param {boolean} [hasJSActionsPromise]
    * @returns {AnnotationLayerBuilder}
    */
   createAnnotationLayerBuilder(
@@ -145,7 +149,7 @@ class DefaultAnnotationLayerFactory {
     renderInteractiveForms = true,
     l10n = NullL10n,
     enableScripting = false,
-    hasJSActions = false
+    hasJSActionsPromise = null
   ) {
     return new AnnotationLayerBuilder({
       pageDiv,
@@ -156,7 +160,7 @@ class DefaultAnnotationLayerFactory {
       l10n,
       annotationStorage,
       enableScripting,
-      hasJSActions,
+      hasJSActionsPromise,
     });
   }
 }
diff --git a/web/base_viewer.js b/web/base_viewer.js
index cf2c8552a..966eaf342 100644
--- a/web/base_viewer.js
+++ b/web/base_viewer.js
@@ -468,7 +468,6 @@ class BaseViewer {
 
     const annotationStorage = pdfDocument.annotationStorage;
     const optionalContentConfigPromise = pdfDocument.getOptionalContentConfig();
-    const hasJSActionsPromise = pdfDocument.hasJSActions();
 
     this._pagesCapability.promise.then(() => {
       this.eventBus.dispatch("pagesloaded", {
@@ -532,7 +531,6 @@ class BaseViewer {
             maxCanvasPixels: this.maxCanvasPixels,
             l10n: this.l10n,
             enableScripting: this.enableScripting,
-            hasJSActionsPromise,
           });
           this._pages.push(pageView);
         }
@@ -1214,8 +1212,7 @@ class BaseViewer {
    *   for annotation icons. Include trailing slash.
    * @param {boolean} renderInteractiveForms
    * @param {IL10n} l10n
-   * @param {boolean} enableScripting
-   * @param {boolean} hasJSActions
+   * @param {boolean} [enableScripting]
    * @returns {AnnotationLayerBuilder}
    */
   createAnnotationLayerBuilder(
@@ -1225,8 +1222,7 @@ class BaseViewer {
     imageResourcesPath = "",
     renderInteractiveForms = false,
     l10n = NullL10n,
-    enableScripting = false,
-    hasJSActions = false
+    enableScripting = false
   ) {
     return new AnnotationLayerBuilder({
       pageDiv,
@@ -1238,7 +1234,7 @@ class BaseViewer {
       downloadManager: this.downloadManager,
       l10n,
       enableScripting,
-      hasJSActions,
+      hasJSActionsPromise: enableScripting && this.pdfDocument.hasJSActions(),
     });
   }
 
diff --git a/web/interfaces.js b/web/interfaces.js
index 062369362..41a4fe571 100644
--- a/web/interfaces.js
+++ b/web/interfaces.js
@@ -181,8 +181,8 @@ class IPDFAnnotationLayerFactory {
    *   for annotation icons. Include trailing slash.
    * @param {boolean} renderInteractiveForms
    * @param {IL10n} l10n
-   * @param {boolean} enableScripting
-   * @param {boolean} hasJSActions
+   * @param {boolean} [enableScripting]
+   * @param {boolean} [hasJSActionsPromise]
    * @returns {AnnotationLayerBuilder}
    */
   createAnnotationLayerBuilder(
@@ -193,7 +193,7 @@ class IPDFAnnotationLayerFactory {
     renderInteractiveForms = true,
     l10n = undefined,
     enableScripting = false,
-    hasJSActions = false
+    hasJSActionsPromise = null
   ) {}
 }
 
diff --git a/web/pdf_page_view.js b/web/pdf_page_view.js
index 2d28d9d8c..72b338764 100644
--- a/web/pdf_page_view.js
+++ b/web/pdf_page_view.js
@@ -63,10 +63,8 @@ import { viewerCompatibilityParams } from "./viewer_compatibility.js";
  *   total pixels, i.e. width * height. Use -1 for no limit. The default value
  *   is 4096 * 4096 (16 mega-pixels).
  * @property {IL10n} l10n - Localization service.
- * @property {boolean} enableScripting - Enable embedded script execution.
+ * @property {boolean} [enableScripting] - Enable embedded script execution.
  *   The default value is `false`.
- * @property {Promise<boolean>} hasJSActionsPromise - A promise that is resolved
- * with true if some fields have some JS actions.
  */
 
 const MAX_CANVAS_PIXELS = viewerCompatibilityParams.maxCanvasPixels || 16777216;
@@ -114,7 +112,6 @@ class PDFPageView {
     this.enableWebGL = options.enableWebGL || false;
     this.l10n = options.l10n || NullL10n;
     this.enableScripting = options.enableScripting || false;
-    this.hasJSActionsPromise = options.hasJSActionsPromise;
 
     this.paintTask = null;
     this.paintedViewportMap = new WeakMap();
@@ -547,30 +544,27 @@ class PDFPageView {
       }
     );
 
-    return this.hasJSActionsPromise.then(hasJSActions => {
-      if (this.annotationLayerFactory) {
-        if (!this.annotationLayer) {
-          this.annotationLayer = this.annotationLayerFactory.createAnnotationLayerBuilder(
-            div,
-            pdfPage,
-            this._annotationStorage,
-            this.imageResourcesPath,
-            this.renderInteractiveForms,
-            this.l10n,
-            this.enableScripting,
-            hasJSActions
-          );
-        }
-        this._renderAnnotationLayer();
+    if (this.annotationLayerFactory) {
+      if (!this.annotationLayer) {
+        this.annotationLayer = this.annotationLayerFactory.createAnnotationLayerBuilder(
+          div,
+          pdfPage,
+          this._annotationStorage,
+          this.imageResourcesPath,
+          this.renderInteractiveForms,
+          this.l10n,
+          this.enableScripting
+        );
       }
-      div.setAttribute("data-loaded", true);
+      this._renderAnnotationLayer();
+    }
+    div.setAttribute("data-loaded", true);
 
-      this.eventBus.dispatch("pagerender", {
-        source: this,
-        pageNumber: this.id,
-      });
-      return resultPromise;
+    this.eventBus.dispatch("pagerender", {
+      source: this,
+      pageNumber: this.id,
     });
+    return resultPromise;
   }
 
   paintOnCanvas(canvasWrapper) {
```

Reply:
Oh, I'm perfectly fine with this approach of delaying `hasJSActions`. My concerns here were only with the changes in `src/core/annotation.js` since it doesn't feel right to add a document-wide property on annotation instances (not the boolean itself being the problem, but merely the location of the data being out of place). Annotation instances should only contain information about the annotations themselves, and `hasJSActions` is not a property of the annotation but of the document. A solution along the lines of `enableScripting`, which also isn't included in every annotation instance, was merely a direction that could be thought about to move that data out of the annotation instances, but that wasn't really clear from my side; sorry for that.